### PR TITLE
mbim.sh: use umbim config instead of dhcp client

### PR DESCRIPTION
**Pull request by jcadduono**
I noticed after purchasing my Sierra EM7565 and EM7455 that there were a few issues:

1. Sierra has a weird PIN status return for a non-existent pin2. I figured this might be related to the EM7xxx dual sim capability, but upon disabling dual sim capability through AT commands I still got the error. Thus, I have added the ignore_pinfail config option.
2. The mbim connection initialization actually retrieves the IP addresses for us, so DHCP is not required. (thank you @bmork for this revelation lol)
3. Because of 2, I was failing to get an IP address for IPv4. (IP was already obtained, but not being used) It was still working for IPv6 because of IPv6 magic however. (seriously, cool, I can set a static IPv6 address for pretty much anything in my providers IPv6 gateway subnet and it will work, wow)

Okay so there's some considerations to be had here...
I've added pdptype to mbim.sh because umbim (secretly) supports it, however, it does not appear to support it very well as upon testing, the umbim client will just get stuck when using ipv4 or ipv6 only pdptype. I know that my mobile provider supports all of these pdptypes, so maybe as a temporary solution we could simply use pdptype variable to decide if we want to ignore an IP version reported by umbim config and always assume that umbim will connect with ipv4v6.

Regarding the pin fail ignoring, perhaps that is another issue left to umbim? I don't know the protocol at all but maybe a flag is getting misinterpreted somewhere? Unlikely, I suppose.

I wrote a luci-proto-mbim module and would be happy to submit that after this PR is looked over.

I am not sure if it's good or bad to set the mtu of wwan0 to 1460, as maybe mtu size is already being handled by the modem hardware?

Major issue: Clients on the network can't seem to be able to access the outside world on IPv6?
They are receiving DHCPv6 addresses and default gateway to br-lan, as well as DNS servers which are working!
Maybe a firewall issue odhp6c somehow escapes?
I'm stuck on this one and don't know how to fix it yet.
I'm hoping you guys would have some answers.

@bmork @blogic

### CODE REVIEW

**Comment by obsy** [comment]
In the same way we can configure qmi proto: https://forum.openwrt.org/t/troubleshooting-qmi-connection-to-a-lte-modem-does-not-get-ip-address/14926/21

**Comment by bmork** [comment]
Cezary Jackiewicz <notifications@github.com> writes:

> In the same way we can configure qmi proto:
> https://forum.openwrt.org/t/troubleshooting-qmi-connection-to-a-lte-modem-does-not-get-ip-address/14926/21

Yes, the issue is the same.  DHCP is an optional modem firmware feature.
We must fall back to the QMI/MBIM IP config requests if DHCP is
unsupported.  Which makes DHCP pointless in the first place. Much better
to just set up a static interface.

While doing this: Note that gateway address and netmask values are
generated by the modem firmware.  The real mobile connection is a
point-to-point link with a single local /32 and an unnumbered remote.
You may use the values provided by the modem, or you can let the proto
script figure it out.  Both values are obviously pointless on MBIM and
raw QMI, since there are no L2 headers on the USB link.  I.e. there is
nothing to ARP for.



Bjørn

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh [hunk 1]
```diff
@@ -84,93 +89,143 @@ _proto_mbim_setup() {
 	echo "mbim[$$]" "Checking pin"
 	umbim $DBG -n -t $tid -d $device pinstate || {
 		echo "mbim[$$]" "PIN required"
-		proto_notify_error "$interface" PIN_FAILED
-		proto_block_restart "$interface"
-		return 1
+		if [ ! "$ignore_pinfail" ] || [ "$ignore_pinfail" = 0 ]; then
```
Comment (bmork) [inline package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: I don't think you should make this a config option.  It's just unnecessary hassle when configuring these modems.  And worse: You do not want to ignore PIN1 failures.

Parse the "required pin:" value and DTRT depening on what it is.

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh [hunk 2]
```diff
@@ -84,93 +89,143 @@ _proto_mbim_setup() {
...
+			return 1
+		else
+			echo "mbim[$$]" "Ignoring PIN failure"
+		fi
 	}
 	tid=$((tid + 1))
 
 	echo "mbim[$$]" "Checking subscriber"
- 	umbim $DBG -n -t $tid -d $device subscriber || {
+	umbim $DBG -n -t $tid -d $device subscriber || {
```
Comment (bmork) [inline package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: Whitespace cleanup is nice, but reviewing is simpler if that's kept in a separate patch

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh [hunk 3]
```diff
@@ -84,93 +89,143 @@ _proto_mbim_setup() {
...
+	_proto_mbim_setup "$@"
 	ret=$?
 
 	[ "$ret" = 0 ] || {
 		logger "mbim bringup failed, retry in 15s"
 		sleep 15
 	}
 
-	return $rt
+	return $ret
```
Comment (bmork) [inline package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: Unrelated fixes should also be split out in separate patches IMHO

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh [hunk 4]
```diff
@@ -84,93 +89,143 @@ _proto_mbim_setup() {
...
+		proto_add_ipv6_address "$ip" "$mask"
+		proto_add_ipv6_prefix "$ip_6"
+		[ "$defaultroute" = 0  ] || proto_add_ipv6_route "::" 0 "$gateway" "" "" "$ip_6" "$metric"
+		[ "$peerdns" = 0 ] || {
+			[ "$dns1" ] && proto_add_dns_server "$dns1"
+			[ "$dns2" ] && proto_add_dns_server "$dns2"
+		}
+		[ "$mtu" ] && ip link set "$ifname" mtu "$mtu"
+		# TODO: Don't know how to propagate IPv6 address to lan from here?
+	}
```
Comment (bmork) [inline package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: This looks nice to me, but I am no expert in netifd or OpenWrt. 

I guess you would want to optionally plug the IPv6 prefix into odhcpd here too?

**Review by bmork — CHANGES_REQUESTED** [review]

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh [hunk 5]
```diff
@@ -84,93 +89,143 @@ _proto_mbim_setup() {
...
 	echo "mbim[$$]" "Connect to network"
-	while ! umbim $DBG -n -t $tid -d $device connect "$apn" "$auth" "$username" "$password"; do
+	while ! umbim $DBG -n -t $tid -d $device connect "$pdptype:$apn" "$auth" "$username" "$password"; do
 		tid=$((tid + 1))
 		sleep 1;
 	done
 	tid=$((tid + 1))
 
+	echo "mbim[$$]" "Connected, obtain IP address and configure interface"
+	config="/tmp/mbim.$$.config"
```
Comment (jow-) [inline package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: I think /var/run might be a more appropriate place.

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh [hunk 6]
```diff
@@ -84,93 +89,143 @@ _proto_mbim_setup() {
...
+		return 1
+	}
+	cat "$config"
+	tid=$((tid + 1))
+
 	uci_set_state network $interface tid "$tid"
 
-	echo "mbim[$$]" "Connected, starting DHCP"
+	local ip_4 ip_6
+	ip_4=$(awk '$1=="ipv4address:" {print $2}' "$config")
```
Comment (jow-) [inline package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: I'd prefer `sed -ne 's/^ipv4address: *//p "$config"`

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh [hunk 7]
```diff
@@ -84,93 +89,143 @@ _proto_mbim_setup() {
...
+	proto_set_keep 1
+	local ip mask gateway mtu dns1 dns2
+
+	[ "$ip_4" ] && {
+		echo "mbim[$$]" "Configure IPv4 on $ifname"
+		ip=${ip_4%%/*}
+		mask=${ip_4##*/}
+		gateway=$(awk '$1=="ipv4gateway:" {print $2; exit}' "$config")
+		mtu=$(awk '$1=="ipv4mtu:" {print $2; exit}' "$config")
+		dns1=$(awk '$1=="ipv4dnsserver:" {print $2}' "$config" | sed -n "1p")
```
Comment (jow-) [inline package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: `dns_servers=$(sed -ne 's/ipv4dnsserver: *//p' "$config")` and loop over `$dns_servers` below.

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh [hunk 8]
```diff
@@ -84,93 +89,143 @@ _proto_mbim_setup() {
...
+		dns2=$(awk '$1=="ipv4dnsserver:" {print $2}' "$config" | sed -n "2p")
+
+		proto_add_ipv4_address "$ip" "$mask"
+		[ "$defaultroute" = 0  ] || proto_add_ipv4_route 0.0.0.0 0 "$gateway" "$ip_4" "$metric"
```
Comment (jow-) [inline package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: Duplicate whitespace here.

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh [hunk 9]
```diff
@@ -84,93 +89,143 @@ _proto_mbim_setup() {
...
+		ip=${ip_6%%/*}
+		mask=${ip_6##*/}
+		gateway=$(awk '$1=="ipv6gateway:" {print $2; exit}' "$config")
+		mtu=$(awk '$1=="ipv6mtu:" {print $2; exit}' "$config")
+		dns1=$(awk '$1=="ipv6dnsserver:" {print $2}' "$config" | sed -n "1p")
+		dns2=$(awk '$1=="ipv6dnsserver:" {print $2}' "$config" | sed -n "2p")
+
+		proto_add_ipv6_address "$ip" "$mask"
+		proto_add_ipv6_prefix "$ip_6"
+		[ "$defaultroute" = 0  ] || proto_add_ipv6_route "::" 0 "$gateway" "" "" "$ip_6" "$metric"
```
Comment (jow-) [inline package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: Duplicate whitespace here.

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh [hunk 10]
```diff
@@ -84,93 +89,143 @@ _proto_mbim_setup() {
...
+		[ "$peerdns" = 0 ] || {
+			[ "$dns1" ] && proto_add_dns_server "$dns1"
+			[ "$dns2" ] && proto_add_dns_server "$dns2"
+		}
+		[ "$mtu" ] && ip link set "$ifname" mtu "$mtu"
+		# TODO: Don't know how to propagate IPv6 address to lan from here?
```
Comment (jow-) [inline package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: You cannot propagate the interface IPv6, you need a separate routed prefix for that. Not sure if such mobile connections support requesting IPv6 prefixes via DHCPv6-PD.

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh [hunk 11]
```diff
@@ -84,93 +89,143 @@ _proto_mbim_setup() {
...
 }
 
 proto_mbim_teardown() {
 	local interface="$1"
 
-	local device
+	local device tid
 	json_get_vars device
-	local tid=$(uci_get_state network $interface tid)
+	tid=$(uci_get_state network "$interface" tid)
```
Comment (jow-) [inline package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: This is actually unrelated to your PR but the use of uci_get_stage or uci_set_stage should be considered deprecated as the underlying uci state mechanism is deprecated as well. It would be better to store the `tid` in the ubus interface data section. Example of using interface data can be seen here: https://github.com/openwrt/openwrt/blob/master/package/network/config/netifd/files/lib/netifd/dhcp.script#L49

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh (hunk 10 above)
Comment (bmork) [reply package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: DHCPv6-PD is possible, but I've never seen it implemented by any operator.  So I believe we can ignore it for now.

But you can propagate the /64 you are assigned. The mobile interface is not a LAN interface, even if the drivers or modem firmware make it look like one.  There is no on-link prefix.  The assigned /64 is a routed prefix. There is no "interface address".  The modem firmware will just print a random address every time you query it.  You  can safely ignore it.  Steal a single /128 for the OpenWrt router, if you need IPv6 connectivity to it, and route the rest of the /64 wherever you need it.

Notice how the modem firmware gives a new address for every query, despite the connection and prefix being the same (this is the EM7455 in my laptop, but that's irrelevant):


```
root@miraculix:/home/bjorn# umbim -p -d /dev/cdc-wdm0 config 
  ipv4address: 10.197.125.191/25
  ipv4gateway: 10.197.125.192
  ipv4mtu: 1500
  ipv4dnsserver: 193.213.112.4
  ipv4dnsserver: 130.67.15.198
  ipv6address: 2a02:2121:288:ef9d:e832:4b63:ea3:ad82/64
  ipv6gateway: 2a02:2121:288:ef9d:f446:f8f3:886b:97d0
  ipv6mtu: 1500
  ipv6dnsserver: 2001:4600:4:fff:0:0:0:52
  ipv6dnsserver: 2001:4600:4:1fff:0:0:0:52
root@miraculix:/home/bjorn# umbim -p -d /dev/cdc-wdm0 config 
  ipv4address: 10.197.125.191/25
  ipv4gateway: 10.197.125.192
  ipv4mtu: 1500
  ipv4dnsserver: 193.213.112.4
  ipv4dnsserver: 130.67.15.198
  ipv6address: 2a02:2121:288:ef9d:195b:a6ca:6d1d:f9c/64
  ipv6gateway: 2a02:2121:288:ef9d:f446:f8f3:886b:97d0
  ipv6mtu: 1500
  ipv6dnsserver: 2001:4600:4:fff:0:0:0:52
  ipv6dnsserver: 2001:4600:4:1fff:0:0:0:52
root@miraculix:/home/bjorn# umbim -p -d /dev/cdc-wdm0 config 
  ipv4address: 10.197.125.191/25
  ipv4gateway: 10.197.125.192
  ipv4mtu: 1500
  ipv4dnsserver: 193.213.112.4
  ipv4dnsserver: 130.67.15.198
  ipv6address: 2a02:2121:288:ef9d:bc67:381e:7bf8:f241/64
  ipv6gateway: 2a02:2121:288:ef9d:f446:f8f3:886b:97d0
  ipv6mtu: 1500
  ipv6dnsserver: 2001:4600:4:fff:0:0:0:52
  ipv6dnsserver: 2001:4600:4:1fff:0:0:0:52

```

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh (hunk 8 above)
Comment (jcadduono) [reply package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: not a fan of my coding style? i thought it'd be better to have an empty line between proto/json configs and the rest of it, i can remove it though

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh (hunk 6 above)
Comment (jcadduono) [reply package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: would have to remove the ^, the lines start with spaces, awk is really useful for trimming, are you sure you want to make it harder to understand?

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh (hunk 5 above)
Comment (jcadduono) [reply package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: if i went with /var/run do you think it'd be better to store $$ in proto data and rm the file when the connection closes instead of at the end of configuration, or keep as is?

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh (hunk 4 above)
Comment (jcadduono) [reply package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: the comment is wrong, everything is being propagated just fine, i actually can't seem to figure out why clients can't ping the outside world, the routes all look right to me, client configurations look perfect, i can't explain it

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh (hunk 1 above)
Comment (jcadduono) [reply package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: i guess i'll take a look inside umbim cli and see what possible responses there are :/
what is pin2? is it specific to sierra? is it possible pin2 could be a sim pin on a secondary sim on some device that supports it that isn't sierra?
it's hard to know what exactly we should be looking for here... should i also ignore and proceed on "Not supported"?

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh (hunk 6 above)
Comment (jow-) [reply package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: No, if you want to keep awk, its fine.

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh (hunk 8 above)
Comment (jow-) [reply package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: No I mean two consecutive spaces in `... 0  ] ...`, the empty lines are totally fine.

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh (hunk 5 above)
Comment (jow-) [reply package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: I took another look and it appears the config is not needed at all, at least from what I can see. You could either store it in a buffer (`output="$(umbim $DBG -n -t $tid -d $device config)"` and later `ip_4="$(echo "$output" | awk '$1=="ipv4address:" {print $2}')"`).

But its just a suggestion, if you do not want to change it, it is fine with me as well.

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh (hunk 1 above)
Comment (bmork) [reply package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: > what is pin2?

It's a PIN code locking other SIM features than network access.  I tried to Google for the correct definition, but this proved somewhat difficult. The best I found was this: https://www.nttdocomo.co.jp/english/support/trouble/password/pin/

PIN2 is a standard SIM feature. It is not Sierra Wireless specific.  But requesting it when you don't actually use any SIM features requiring it seems to be SWI specific.

I suggest looking specifically for PIN1 unlocking requests and ignoring anything else.  You probably don't want a script like this to handle PUK1 or PUK2 codes :-)

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh (hunk 11 above)
Comment (jcadduono) [reply package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: done, i think

**Comment by jcadduono** [comment]
maybe it would be best to just use this method for ipv4, and continue to use dhcp for ipv6, we can check to see if modem responds with ipv6 address, and if it does, just initiate ipv6 dhcp?

File: package/network/utils/umbim/files/lib/netifd/proto/mbim.sh (hunk 8 above)
Comment (jcadduono) [reply package/network/utils/umbim/files/lib/netifd/proto/mbim.sh]: o am blind

**Comment by bobafetthotmail** [comment]
@jcadduono your last commit lacks signed-off-by.

I'm really interested in your  luci-proto-mbim module too as I do rely on LTE modems.

Also, you can have a look at scripts used in ROOTer firmware (which is a OpenWrt downstream project). https://github.com/danielewood/sierra-wireless-modems/pull/16#issuecomment-503243363

**Comment by jcadduono** [comment]
@bobafetthotmail I have not signed off on the last commit because it does not currently work. IPv6 is not being routed correctly and it will need some additional changes.

**Comment by jcadduono** [comment]
yeah I have no idea how to configure IPv6 other than dhcp so I kept it with ipv6 dhcp + ipv4 static as a local change, I can PR that instead?

**Comment by adrianschmutzler** [comment]
@jow- This is lying around for more than half a year now, and you were the only one commenting on it. Despite, it's marked RFC and the author seems to need feedback.
Would be nice if you could have a quick look and either give him a hint or go for a "Won't implement" and close it, as I don't think anything will happen otherwise.

**Comment by jcadduono** [comment]
i've pushed a commit with the script as i've been using for the past year or so, that has worked for me with no issues whatsoever.

**Comment by adrianschmutzler** [comment]
Your recent commits have formal issues again (missing Signed-off on first sight).

**Comment by jow-** [comment]
Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/jow.git
